### PR TITLE
feature/WMA-5/Login-api

### PR DESCRIPTION
**Pull request by Caleb-Nieto**
### Feature / Bug Description
Adds Login and Reset password services

### Solution
Following dependencies installed: "@nestjs/jwt", "bcryptjs" "class-transformer", "class-validator".  Added Global pipes in main.ts and ".addBearerAuth()" to specify auth in the documentation. 
Generated auth and users resources.
Create userEntity
Login and ResetPasword controllers implemented.
openApi documentation on endpoint /auth/login and /auth/reset-password

### Notes
The resetPassword service is probably breaking some programming rules but is a proof that jwt works

### Tickets
[WMA-5](https://cdentalcaregroup.atlassian.net/browse/WMA-5)

### Screenshots / Screencasts
Swagger
image

Files added
image

[WMA-5]: https://cdentalcaregroup.atlassian.net/browse/WMA-5?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/modules/auth/auth.service.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+export class AuthService {
+  constructor(
+    private readonly jwtService: JwtService,
+    private readonly usersSevice: UsersService,
+  ) {}
+
+  async login(data: LoginDto): Promise<{ access_token: string }> {
+    const user = await this.usersSevice.findOneByEmail(data.email);
+    if (!user) {
+      throw new UnauthorizedException('Email does not exist');
```
Comment (imadiaz) [inline src/modules/auth/auth.service.ts:22]: in the constant file, can you create a String file? and put this string error message there, then we can use it across the app the same string

File: src/modules/auth/dto/login.dto.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { Transform } from 'class-transformer';
+import { IsEmail, IsString, MinLength } from 'class-validator';
+
+export class LoginDto {
```
Comment (imadiaz) [inline src/modules/auth/dto/login.dto.ts:5]: can u update this to LoginDTO i mean use the DTO sufix in upper case

File: src/modules/auth/dto/resetPassword.dto.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { Transform } from 'class-transformer';
+import { IsEmail, IsString, MinLength } from 'class-validator';
+
+export class ResestPasswordDto {
```
Comment (imadiaz) [inline src/modules/auth/dto/resetPassword.dto.ts:5]: same this ResestPasswordDTO

File: src/modules/users/users.service.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    private readonly userRepository: Repository<UserEntity>,
+  ) {}
+
+  findOneByEmail(email: string) {
+    return this.userRepository.findOneBy({ email });
+  }
+  async update(user: UserEntity) {
+    const exists = await this.userRepository.existsBy({ email: user.email });
+    if (!exists) {
+      throw new BadRequestException('The user does not exist');
```
Comment (imadiaz) [inline src/modules/users/users.service.ts:19]: same for this string

File: src/modules/auth/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import * as bcryptjs from 'bcryptjs';
+import { JwtService } from '@nestjs/jwt';
+import { ResestPasswordDto } from './dto/resetPassword.dto';
+import { UsersService } from '../users/users.service';
+
+@Injectable()
+export class AuthService {
+  constructor(
+    private readonly jwtService: JwtService,
+    private readonly usersSevice: UsersService,
```
Comment (imadiaz) [inline src/modules/auth/auth.service.ts:16]: awesome! this is helpful to inject the services!

File: src/modules/users/users.module.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import { Module } from '@nestjs/common';
+import { UsersService } from './users.service';
+import { UsersController } from './users.controller';
+import { TypeOrmModule } from '@nestjs/typeorm';
+import { UserEntity } from './entities/user.entity';
+
+@Module({
+  imports: [TypeOrmModule.forFeature([UserEntity])],
+  controllers: [UsersController],
+  providers: [UsersService],
```
Comment (imadiaz) [inline src/modules/users/users.module.ts:10]: if we just use the service, i think (not for now) we can delete the UserController and the UserModule in the future and just keep the UserService, this change is not part of this scope but just to keep in mind

File: src/modules/auth/dto/login.dto.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { ApiProperty } from '@nestjs/swagger';
```
Comment (imadiaz) [inline src/modules/auth/dto/login.dto.ts:1]: I think that you should move this file into this folder, the correct path should be:

`/src/modules/auth/models/dto/login.dto.ts`

just add "models" folder

**Review by imadiaz — COMMENTED** [review]
just some comments and improvements but looks good!

File: src/modules/auth/auth.service.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    const payload = { email: user.email };
+    const access_token = await this.jwtService.signAsync(payload);
+
+    return {
+      access_token,
+    };
+  }
+  async resetPassword(data: ResestPasswordDto, email: string) {
```
Comment (imadiaz) [inline src/modules/auth/auth.service.ts:40]: on my personal side, i really like to have the methods like this:
`resetPassword = async (data: ResestPasswordDto, email: string) => {
    
   }`

this is just an opinion if you want to follow this or keep your implementation!

File: src/modules/auth/dto/resetPassword.dto.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+  @ApiProperty({ description: 'password', example: '********', minimum: 8 })
+  @Transform(({ value }) => value.trim())
+  @IsString()
+  @MinLength(8)
+  newPassword;
```
Comment (imadiaz) [inline src/modules/auth/dto/resetPassword.dto.ts:10]: you can also add the email in this DTO

File: src/modules/auth/dto/login.dto.ts (hunk 2 above)
Comment (Caleb-Nieto) [reply src/modules/auth/dto/login.dto.ts:5]: Right now

File: src/modules/auth/dto/resetPassword.dto.ts (hunk 3 above)
Comment (Caleb-Nieto) [reply src/modules/auth/dto/resetPassword.dto.ts:5]: Right now

File: src/modules/auth/auth.service.ts (hunk 5 above)
Comment (Caleb-Nieto) [reply src/modules/auth/auth.service.ts:16]: Got it thank u for all the feedback

File: src/modules/users/users.module.ts (hunk 6 above)
Comment (Caleb-Nieto) [reply src/modules/users/users.module.ts:10]: Okay understood not for now!!

File: src/modules/auth/dto/login.dto.ts (hunk 7 above)
Comment (Caleb-Nieto) [reply src/modules/auth/dto/login.dto.ts:1]: done!

File: src/modules/auth/auth.service.ts (hunk 8 above)
Comment (Caleb-Nieto) [reply src/modules/auth/auth.service.ts:40]: Sure I ll keep it in mind here we go arrow functions

File: src/modules/auth/dto/resetPassword.dto.ts (hunk 9 above)
Comment (Caleb-Nieto) [reply src/modules/auth/dto/resetPassword.dto.ts:10]: done

File: src/modules/users/users.service.ts (hunk 4 above)
Comment (Caleb-Nieto) [reply src/modules/users/users.service.ts:19]: Im goint to try I hope I understood

**Review by imadiaz — APPROVED** [review]

**Pull request merged**